### PR TITLE
[Attention] Add out dtype for QK matmul

### PR DESCRIPTION
**Pull request by CharlieFRuan**
This PR adds an argument ~`attn_weights_out_dtype`~ `qk_dtype` for `attention.py`, essentially making `matmul(q, k, out_dtype=qk_dtype)`.

The motivation is that some models (e.g. phi), when operating under float16, can observe INF in Q@K. Thus, we accumulate in float32 to avoid this.

### CODE REVIEW

File: python/mlc_chat/op/attention.py [hunk 1]
```diff
@@ -16,12 +16,13 @@
...
 
 
-def attention(  # pylint: disable=invalid-name,too-many-locals,too-many-statements
+def attention(  # pylint: disable=invalid-name,too-many-locals,too-many-statements,too-many-arguments
     q: nn.Tensor,
     k: nn.Tensor,
     v: nn.Tensor,
     casual_mask: nn.Tensor,
     attn_score_scaling_factor: float = 1.0,
+    attn_weights_out_dtype: str = None,
```
Comment (junrushao) [inline python/mlc_chat/op/attention.py:25]: I'm a bit skeptical of what `attn_weights` refers to given it can mean many things. How about `qk_dtype`?

Comment (CharlieFRuan) [reply python/mlc_chat/op/attention.py]: Makes sense, fixed; thank you!

**Comment by junrushao** [comment]
In FlashInfer, there’s a parameter called “fp16_qk”, which we could tweak accordingly as well 😆

**Comment by CharlieFRuan** [comment]
Ahh thanks for the pointer, I'll try and test a bit

**Comment by tqchen** [comment]
also cc @jinhongyii @spectrometerHBH @MasterJH5574 , we likely will also need to pay attention to this in TIR KVCache kernels

**Comment by CharlieFRuan** [comment]
Updated `fp16_qk` for FlashInfer; tested on llama

**Review by junrushao — APPROVED** [review]
Great work! Also CC @LeshengJin as he's digging precision issues in Phi

**Pull request merged**